test/e2e/clusterctl_upgrade.go

```diff
@@ -811,7 +852,62 @@ func calculateExpectedWorkerCount(ctx context.Context, c client.Client, unstruct
 		if !ok {
 			continue
 		}
-		expectedWorkerCount += replicas
+		expectedMachineDeploymentWorkerCount += replicas
+	}
+
+	return expectedMachineDeploymentWorkerCount
+}
+
+func calculateExpectedMachinePoolMachineCount(ctx context.Context, c client.Client, workloadClusterNamespace, workloadClusterName string) int64 {
+	expectedMachinePoolMachineCount := int64(0)
+
+	machinePoolList := &expv1.MachinePoolList{}
+	if err := c.List(
+		ctx,
+		machinePoolList,
+		client.InNamespace(workloadClusterNamespace),
+		client.MatchingLabels{clusterv1.ClusterNameLabel: workloadClusterName},
+	); err == nil {
+		for _, mp := range machinePoolList.Items {
+			infraMachinePool := getInfraMachinePool(ctx, c, mp, workloadClusterNamespace, workloadClusterName)
+			if infraMachinePool != nil {
+				// Check if the InfraMachinePool has an infrastructureMachineKind field. If it does not, we should skip checking for MachinePool machines.
+				err := util.UnstructuredUnmarshalField(infraMachinePool, ptr.To(""), "status", "infrastructureMachineKind")
```

Comment:
I think we should not ignore errors that are not  ErrUnstructuredFieldNotFound

Reply:
Seems wrong to add mp.Status.ReadyReplicas if we got an unknown error

Also, is expected really the number of currently readyReplicas? Should this be mp.Spec.Replicas instead?

Reply:
I think overall this should look like this
```go
				err := util.UnstructuredUnmarshalField(infraMachinePool, ptr.To(""), "status", "infrastructureMachineKind")
				if err != nil && !errors.Is(err, util.ErrUnstructuredFieldNotFound) {
					return 0, err
				}
				if err == nil {
					expectedMachinePoolMachineCount += int64(mp.Spec.Replicas)
				}
```